File: src/causal_inference/bayesian/bayesian_propensity.py
"""
Bayesian Propensity Score Estimation.

Implements two approaches:
1. Stratified Beta-Binomial (conjugate, for discrete covariates)
2. Bayesian Logistic Regression (Laplace approximation, for continuous covariates)

Session 102: Initial implementation.

Mathematical Foundation
-----------------------
**Stratified Beta-Binomial**:

    Prior:      e_s ~ Beta(α₀, β₀) for each stratum s
    Likelihood: T_i | s ~ Bernoulli(e_s)
    Posterior:  e_s | T ~ Beta(α₀ + n_treated_s, β₀ + n_control_s)

**Bayesian Logistic Regression** (Laplace approximation):

    Prior:      β ~ N(0, σ²_prior I)
    Likelihood: T | X ~ Bernoulli(sigmoid(Xβ))
    Posterior:  β | T, X ≈ N(β_MLE, H⁻¹)

where H is the Hessian of the negative log-posterior at β_MLE.
"""

from typing import Optional, Tuple, Union

import numpy as np
from numpy.typing import NDArray
from scipy import stats
from scipy.optimize import minimize
from scipy.special import expit  # sigmoid function

from causal_inference.bayesian.types import (
    BayesianPropensityResult,
    StratumInfo,
)


def _create_strata(
    covariates: NDArray[np.float64],
    n_bins: int = 5,
) -> Tuple[NDArray[np.int64], int]:
    """
    Create strata from covariates by discretizing continuous variables.

    Parameters
    ----------
    covariates : ndarray
        Covariate matrix, shape (n, p).
    n_bins : int
        Number of bins for each continuous covariate.

    Returns
    -------
    strata : ndarray
        Stratum assignment for each observation, shape (n,).
    n_strata : int
        Total number of unique strata.
    """
    n, p = covariates.shape

    # Discretize each covariate
    discretized = np.zeros((n, p), dtype=np.int64)
    for j in range(p):
        col = covariates[:, j]
        n_unique = len(np.unique(col))

        if n_unique <= n_bins:
            # Already discrete, use as-is
            _, discretized[:, j] = np.unique(col, return_inverse=True)
        else:
            # Discretize using quantile bins
            percentiles = np.linspace(0, 100, n_bins + 1)
            bins = np.percentile(col, percentiles)
            bins[0] = -np.inf
            bins[-1] = np.inf
            discretized[:, j] = np.digitize(col, bins[1:-1])

    # Create unique stratum for each combination
    # Use tuple hashing approach
    strata = np.zeros(n, dtype=np.int64)
    stratum_map = {}
    for i in range(n):
        key = tuple(discretized[i, :])
        if key not in stratum_map:
            stratum_map[key] = len(stratum_map)
        strata[i] = stratum_map[key]

    return strata, len(stratum_map)


def bayesian_propensity_stratified(
    treatment: NDArray[np.float64],
    covariates: NDArray[np.float64],
    prior_alpha: float = 1.0,
    prior_beta: float = 1.0,
    n_bins: int = 5,
    n_posterior_samples: int = 1000,
) -> BayesianPropensityResult:
    """
    Bayesian propensity score estimation using stratified Beta-Binomial.

    For each stratum defined by covariate combinations, estimates propensity
    using Beta-Binomial conjugate model.

    Parameters
    ----------
    treatment : ndarray
        Binary treatment indicator (0/1), shape (n,).
    covariates : ndarray
        Covariate matrix, shape (n, p).
    prior_alpha : float, default=1.0
        Beta prior alpha parameter (prior successes + 1).
    prior_beta : float, default=1.0
        Beta prior beta parameter (prior failures + 1).
        Default (1,1) is uniform prior.
    n_bins : int, default=5
        Number of bins for discretizing continuous covariates.
    n_posterior_samples : int, default=1000
        Number of posterior samples to draw.

    Returns
    -------
    BayesianPropensityResult
        Posterior propensity samples, means, and stratum information.

    Notes
    -----
    The Beta(1,1) prior is uniform on [0,1], representing no prior information.
    Beta(0.5, 0.5) is Jeffreys prior, which is weakly informative.

    Examples
    --------
    >>> import numpy as np
    >>> np.random.seed(42)
    >>> n = 200
    >>> X = np.random.randn(n, 2)
    >>> T = (X[:, 0] > 0).astype(float)
    >>> result = bayesian_propensity_stratified(T, X)
    >>> print(f"Mean propensity: {result['posterior_mean'].mean():.3f}")
    """
    # Input validation
    treatment = np.asarray(treatment, dtype=np.float64)
    covariates = np.asarray(covariates, dtype=np.float64)

    if covariates.ndim == 1:
        covariates = covariates.reshape(-1, 1)

    n = len(treatment)

    if len(covariates) != n:
        raise ValueError(
            f"Length mismatch: treatment ({n}) != covariates ({len(covariates)})"
        )

    if not np.all(np.isin(treatment, [0, 1])):
        raise ValueError("Treatment must be binary (0 or 1)")

    if prior_alpha <= 0 or prior_beta <= 0:
        raise ValueError(
            f"prior_alpha and prior_beta must be positive, "
            f"got alpha={prior_alpha}, beta={prior_beta}"
        )

    # Create strata
    strata, n_strata = _create_strata(covariates, n_bins)

    # Compute posterior for each stratum
    stratum_info = []
    posterior_samples = np.zeros((n_posterior_samples, n))

    for s in range(n_strata):
        mask = strata == s
        n_in_stratum = np.sum(mask)
        n_treated = np.sum(treatment[mask])
        n_control = n_in_stratum - n_treated

        # Posterior parameters
        post_alpha = prior_alpha + n_treated
        post_beta = prior_beta + n_control

        # Posterior mean
        post_mean = post_alpha / (post_alpha + post_beta)

        # Draw samples from posterior
        samples = np.random.beta(post_alpha, post_beta, size=n_posterior_samples)

        # Assign to all observations in this stratum
        posterior_samples[:, mask] = samples[:, np.newaxis]

        stratum_info.append(StratumInfo(
            stratum_id=s,
            n_obs=int(n_in_stratum),
            n_treated=int(n_treated),
            n_control=int(n_control),
            posterior_alpha=post_alpha,
            posterior_beta=post_beta,
            posterior_mean=post_mean,
            posterior_sd=np.sqrt(
                (post_alpha * post_beta) /
                ((post_alpha + post_beta) ** 2 * (post_alpha + post_beta + 1))
            ),
        ))

    # Compute summary statistics
    posterior_mean = np.mean(posterior_samples, axis=0)
    posterior_sd = np.std(posterior_samples, axis=0)

    # Uncertainty metrics
    mean_uncertainty = np.mean(posterior_sd)
    propensity_range = np.ptp(posterior_mean)

    return BayesianPropensityResult(
        posterior_samples=posterior_samples,
        posterior_mean=posterior_mean,
        posterior_sd=posterior_sd,
        strata=strata,
        n_strata=n_strata,
        stratum_info=stratum_info,
        prior_alpha=prior_alpha,
        prior_beta=prior_beta,
        method="stratified_beta_binomial",
        n=n,
        n_treated=int(np.sum(treatment)),
        n_control=int(n - np.sum(treatment)),
        mean_uncertainty=mean_uncertainty,
        propensity_range=propensity_range,
    )


def _neg_log_posterior_logistic(
    beta: NDArray[np.float64],
    X: NDArray[np.float64],
    T: NDArray[np.float64],
    prior_var: float,
) -> float:
    """Negative log-posterior for logistic regression."""
    linear = X @ beta
    # Clip for numerical stability
    linear = np.clip(linear, -500, 500)
    prob = expit(linear)

    # Log-likelihood
    ll = np.sum(T * np.log(prob + 1e-10) + (1 - T) * np.log(1 - prob + 1e-10))

    # Log-prior (normal)
    log_prior = -0.5 * np.sum(beta**2) / prior_var

    return -(ll + log_prior)


def _hessian_logistic(
    beta: NDArray[np.float64],
    X: NDArray[np.float64],
    prior_var: float,
) -> NDArray[np.float64]:
    """Hessian of negative log-posterior for logistic regression."""
    linear = X @ beta
    linear = np.clip(linear, -500, 500)
    prob = expit(linear)

    # Weight matrix
    W = prob * (1 - prob)

    # Hessian = X' diag(W) X + I/prior_var
    hessian = X.T @ (W[:, np.newaxis] * X) + np.eye(X.shape[1]) / prior_var

    return hessian


def bayesian_propensity_logistic(
    treatment: NDArray[np.float64],
    covariates: NDArray[np.float64],
    prior_sd: float = 10.0,
    n_posterior_samples: int = 1000,
    include_intercept: bool = True,
) -> BayesianPropensityResult:
    """
    Bayesian propensity score estimation using logistic regression with Laplace approximation.

    Uses normal approximation to posterior of logistic regression coefficients.

    Parameters
    ----------
    treatment : ndarray
        Binary treatment indicator (0/1), shape (n,).
    covariates : ndarray
        Covariate matrix, shape (n, p).
    prior_sd : float, default=10.0
        Prior standard deviation for regression coefficients.
        Larger values = weaker prior.
    n_posterior_samples : int, default=1000
        Number of posterior samples to draw.
    include_intercept : bool, default=True
        Whether to include an intercept term.

    Returns
    -------
    BayesianPropensityResult
        Posterior propensity samples, means, and coefficient information.

    Notes
    -----
    The Laplace approximation uses a normal distribution centered at the
    maximum a posteriori (MAP) estimate with covariance given by the
    inverse Hessian of the negative log-posterior.

    This approximation is accurate when:
    - Sample size is reasonably large
    - Posterior is approximately normal (unimodal, symmetric)

    Examples
    --------
    >>> import numpy as np
    >>> np.random.seed(42)
    >>> n = 200
    >>> X = np.random.randn(n, 2)
    >>> logit = 0.5 * X[:, 0] + 0.3 * X[:, 1]
    >>> T = (np.random.rand(n) < 1/(1 + np.exp(-logit))).astype(float)
    >>> result = bayesian_propensity_logistic(T, X)
    >>> print(f"Coefficient posterior means: {result['coefficient_mean']}")
    """
    # Input validation
    treatment = np.asarray(treatment, dtype=np.float64)
    covariates = np.asarray(covariates, dtype=np.float64)

    if covariates.ndim == 1:
        covariates = covariates.reshape(-1, 1)

    n, p = covariates.shape

    if len(treatment) != n:
        raise ValueError(
            f"Length mismatch: treatment ({len(treatment)}) != covariates ({n})"
        )

    if not np.all(np.isin(treatment, [0, 1])):
        raise ValueError("Treatment must be binary (0 or 1)")

    if prior_sd <= 0:
        raise ValueError(f"prior_sd must be positive, got {prior_sd}")

    # Add intercept if requested
    if include_intercept:
        X = np.column_stack([np.ones(n), covariates])
    else:
        X = covariates

    n_coef = X.shape[1]
    prior_var = prior_sd**2

    # Find MAP estimate
    beta_init = np.zeros(n_coef)
    result = minimize(
        _neg_log_posterior_logistic,
        beta_init,
        args=(X, treatment, prior_var),
        method="BFGS",
    )
    beta_map = result.x

    # Compute Hessian at MAP for Laplace approximation
    hessian = _hessian_logistic(beta_map, X, prior_var)
    try:
        cov_matrix = np.linalg.inv(hessian)
    except np.linalg.LinAlgError:
        # Use pseudo-inverse if singular
        cov_matrix = np.linalg.pinv(hessian)

    # Draw coefficient samples from approximate posterior
    try:
        beta_samples = np.random.multivariate_normal(
            beta_map, cov_matrix, size=n_posterior_samples
        )
    except np.linalg.LinAlgError:
        # Fallback to independent normals if covariance is problematic
        stds = np.sqrt(np.diag(cov_matrix))
        beta_samples = np.random.normal(
            beta_map, stds, size=(n_posterior_samples, n_coef)
        )

    # Compute propensity samples
    linear_samples = beta_samples @ X.T  # (n_samples, n)
    propensity_samples = expit(linear_samples)

    # Summary statistics
    posterior_mean = np.mean(propensity_samples, axis=0)
    posterior_sd = np.std(propensity_samples, axis=0)

    # Coefficient summaries
    coefficient_mean = np.mean(beta_samples, axis=0)
    coefficient_sd = np.std(beta_samples, axis=0)

    # Uncertainty metrics
    mean_uncertainty = np.mean(posterior_sd)
    propensity_range = np.ptp(posterior_mean)

    return BayesianPropensityResult(
        posterior_samples=propensity_samples,
        posterior_mean=posterior_mean,
        posterior_sd=posterior_sd,
        strata=None,
        n_strata=0,
        stratum_info=None,
        prior_alpha=0.0,
        prior_beta=0.0,
        method="logistic_laplace",
        n=n,
        n_treated=int(np.sum(treatment)),
        n_control=int(n - np.sum(treatment)),
        mean_uncertainty=mean_uncertainty,
        propensity_range=propensity_range,
        coefficient_mean=coefficient_mean,
        coefficient_sd=coefficient_sd,
        coefficient_samples=beta_samples,
        prior_sd=prior_sd,
    )


def bayesian_propensity(
    treatment: NDArray[np.float64],
    covariates: NDArray[np.float64],
    method: str = "auto",
    **kwargs,
) -> BayesianPropensityResult:
    """
    Bayesian propensity score estimation.

    Automatically selects between stratified Beta-Binomial and logistic
    regression based on covariate characteristics.

    Parameters
    ----------
    treatment : ndarray
        Binary treatment indicator (0/1), shape (n,).
    covariates : ndarray
        Covariate matrix, shape (n, p).
    method : str, default="auto"
        Estimation method:
        - "auto": Choose based on covariate characteristics
        - "stratified": Beta-Binomial stratified estimation
        - "logistic": Logistic regression with Laplace approximation
    **kwargs
        Additional arguments passed to the specific method.

    Returns
    -------
    BayesianPropensityResult
        Posterior propensity samples and summary statistics.
    """
    covariates = np.asarray(covariates, dtype=np.float64)
    if covariates.ndim == 1:
        covariates = covariates.reshape(-1, 1)

    if method == "auto":
        # Heuristic: use stratified if few unique values per covariate
        n_unique = [len(np.unique(covariates[:, j])) for j in range(covariates.shape[1])]
        if all(nu <= 10 for nu in n_unique):
            method = "stratified"
        else:
            method = "logistic"

    # Filter kwargs to only include parameters accepted by the selected method
    stratified_params = {"prior_alpha", "prior_beta", "n_bins", "n_posterior_samples"}
    logistic_params = {"prior_sd", "n_posterior_samples", "include_intercept"}

    if method == "stratified":
        filtered_kwargs = {k: v for k, v in kwargs.items() if k in stratified_params}
        return bayesian_propensity_stratified(treatment, covariates, **filtered_kwargs)
    elif method == "logistic":
        filtered_kwargs = {k: v for k, v in kwargs.items() if k in logistic_params}
        return bayesian_propensity_logistic(treatment, covariates, **filtered_kwargs)
    else:
        raise ValueError(f"Unknown method: {method}. Use 'auto', 'stratified', or 'logistic'.")

File: julia/src/bayesian/bayesian_propensity.jl
#=
Bayesian Propensity Score Estimation.

Implements two approaches:
1. Stratified Beta-Binomial (conjugate, for discrete covariates)
2. Bayesian Logistic Regression (Laplace approximation, for continuous covariates)

Session 102: Initial implementation.
=#

# Uses: Distributions, Statistics, LinearAlgebra from main module


"""
    StratumInfo

Information about a propensity stratum in Beta-Binomial estimation.
"""
struct StratumInfo
    stratum_id::Int
    n_obs::Int
    n_treated::Int
    n_control::Int
    posterior_alpha::Float64
    posterior_beta::Float64
    posterior_mean::Float64
    posterior_sd::Float64
end


"""
    BayesianPropensityResult

Result from Bayesian propensity score estimation.
"""
struct BayesianPropensityResult
    posterior_samples::Matrix{Float64}   # (n_samples, n)
    posterior_mean::Vector{Float64}      # (n,)
    posterior_sd::Vector{Float64}        # (n,)
    strata::Union{Vector{Int}, Nothing}
    n_strata::Int
    stratum_info::Union{Vector{StratumInfo}, Nothing}
    prior_alpha::Float64
    prior_beta::Float64
    method::String
    n::Int
    n_treated::Int
    n_control::Int
    mean_uncertainty::Float64
    propensity_range::Float64
    # Logistic method fields
    coefficient_mean::Union{Vector{Float64}, Nothing}
    coefficient_sd::Union{Vector{Float64}, Nothing}
    coefficient_samples::Union{Matrix{Float64}, Nothing}
    prior_sd::Float64
end


"""Pretty-print BayesianPropensityResult."""
function Base.show(io::IO, r::BayesianPropensityResult)
    println(io, "BayesianPropensityResult")
    println(io, "═════════════════════════════════════════")
    println(io, "Method: $(r.method)")
    println(io, "─────────────────────────────────────────")
    println(io, "Sample Information")
    println(io, "  n:         $(r.n)")
    println(io, "  n_treated: $(r.n_treated)")
    println(io, "  n_control: $(r.n_control)")
    if r.n_strata > 0
        println(io, "  n_strata:  $(r.n_strata)")
    end
    println(io, "─────────────────────────────────────────")
    println(io, "Propensity Summary")
    println(io, "  Mean propensity:  $(round(mean(r.posterior_mean), digits=4))")
    println(io, "  Mean uncertainty: $(round(r.mean_uncertainty, digits=4))")
    println(io, "  Range:            $(round(r.propensity_range, digits=4))")
    if r.coefficient_mean !== nothing
        println(io, "─────────────────────────────────────────")
        println(io, "Coefficients")
        for (i, (m, s)) in enumerate(zip(r.coefficient_mean, r.coefficient_sd))
            println(io, "  β[$i]: $(round(m, digits=4)) ± $(round(s, digits=4))")
        end
    end
end


"""
    _create_strata(covariates, n_bins)

Create strata from covariates by discretizing continuous variables.
"""
function _create_strata(
    covariates::AbstractMatrix{<:Real},
    n_bins::Int = 5
)
    n, p = size(covariates)

    # Discretize each covariate
    discretized = zeros(Int, n, p)
    for j in 1:p
        col = covariates[:, j]
        unique_vals = unique(col)

        if length(unique_vals) <= n_bins
            # Already discrete, use as-is
            val_to_idx = Dict(v => i for (i, v) in enumerate(unique_vals))
            discretized[:, j] = [val_to_idx[v] for v in col]
        else
            # Discretize using quantile bins
            percentiles = range(0, 100, length=n_bins+1)
            bins = quantile(col, percentiles ./ 100)
            bins[1] = -Inf
            bins[end] = Inf
            for i in 1:n
                for k in 1:(length(bins)-1)
                    if bins[k] <= col[i] < bins[k+1]
                        discretized[i, j] = k
                        break
                    end
                end
                if col[i] >= bins[end-1]
                    discretized[i, j] = n_bins
                end
            end
        end
    end

    # Create unique stratum for each combination
    strata = zeros(Int, n)
    stratum_map = Dict{Tuple, Int}()
    for i in 1:n
        key = Tuple(discretized[i, :])
        if !haskey(stratum_map, key)
            stratum_map[key] = length(stratum_map) + 1
        end
        strata[i] = stratum_map[key]
    end

    return strata, length(stratum_map)
end


"""
    bayesian_propensity_stratified(treatment, covariates; kwargs...)

Bayesian propensity score estimation using stratified Beta-Binomial.

# Arguments
- `treatment`: Binary treatment indicator (0/1)
- `covariates`: Covariate matrix

# Keyword Arguments
- `prior_alpha=1.0`: Beta prior alpha parameter
- `prior_beta=1.0`: Beta prior beta parameter
- `n_bins=5`: Number of bins for discretization
- `n_posterior_samples=1000`: Number of posterior samples

# Returns
- `BayesianPropensityResult`
"""
function bayesian_propensity_stratified(
    treatment::AbstractVector{<:Real},
    covariates::AbstractMatrix{<:Real};
    prior_alpha::Real = 1.0,
    prior_beta::Real = 1.0,
    n_bins::Int = 5,
    n_posterior_samples::Int = 1000
)
    T = Float64.(treatment)
    X = Float64.(covariates)
    n = length(T)

    if size(X, 1) != n
        throw(ArgumentError(
            "Length mismatch: treatment ($n) != covariates ($(size(X, 1)))"
        ))
    end

    if !all(t -> t == 0 || t == 1, T)
        throw(ArgumentError("Treatment must be binary (0 or 1)"))
    end

    if prior_alpha <= 0 || prior_beta <= 0
        throw(ArgumentError(
            "prior_alpha and prior_beta must be positive, " *
            "got alpha=$prior_alpha, beta=$prior_beta"
        ))
    end

    # Create strata
    strata, n_strata = _create_strata(X, n_bins)

    # Compute posterior for each stratum
    stratum_info = StratumInfo[]
    posterior_samples = zeros(n_posterior_samples, n)

    for s in 1:n_strata
        mask = strata .== s
        n_in_stratum = sum(mask)
        n_treated = Int(sum(T[mask]))
        n_control = n_in_stratum - n_treated

        # Posterior parameters
        post_alpha = prior_alpha + n_treated
        post_beta = prior_beta + n_control

        # Posterior mean
        post_mean = post_alpha / (post_alpha + post_beta)
        post_sd = sqrt(
            (post_alpha * post_beta) /
            ((post_alpha + post_beta)^2 * (post_alpha + post_beta + 1))
        )

        # Draw samples from posterior
        samples = rand(Beta(post_alpha, post_beta), n_posterior_samples)

        # Assign to all observations in this stratum
        for i in 1:n
            if mask[i]
                posterior_samples[:, i] = samples
            end
        end

        push!(stratum_info, StratumInfo(
            s, n_in_stratum, n_treated, n_control,
            post_alpha, post_beta, post_mean, post_sd
        ))
    end

    # Compute summary statistics
    posterior_mean = vec(mean(posterior_samples, dims=1))
    posterior_sd = vec(std(posterior_samples, dims=1))

    mean_uncertainty = mean(posterior_sd)
    propensity_range = maximum(posterior_mean) - minimum(posterior_mean)

    return BayesianPropensityResult(
        posterior_samples,
        posterior_mean,
        posterior_sd,
        strata,
        n_strata,
        stratum_info,
        prior_alpha,
        prior_beta,
        "stratified_beta_binomial",
        n,
        Int(sum(T)),
        Int(n - sum(T)),
        mean_uncertainty,
        propensity_range,
        nothing,
        nothing,
        nothing,
        0.0
    )
end


"""
    _sigmoid(x)

Sigmoid function with numerical stability.
"""
function _sigmoid(x::Real)
    if x >= 0
        z = exp(-x)
        return 1.0 / (1.0 + z)
    else
        z = exp(x)
        return z / (1.0 + z)
    end
end

_sigmoid(x::AbstractArray) = _sigmoid.(x)


"""
    _neg_log_posterior_logistic(beta, X, T, prior_var)

Negative log-posterior for logistic regression.
"""
function _neg_log_posterior_logistic(
    beta::AbstractVector{<:Real},
    X::AbstractMatrix{<:Real},
    T::AbstractVector{<:Real},
    prior_var::Real
)
    linear = X * beta
    linear = clamp.(linear, -500, 500)
    prob = _sigmoid(linear)

    # Log-likelihood
    ll = sum(T .* log.(prob .+ 1e-10) .+ (1 .- T) .* log.(1 .- prob .+ 1e-10))

    # Log-prior (normal)
    log_prior = -0.5 * sum(beta.^2) / prior_var

    return -(ll + log_prior)
end


"""
    _hessian_logistic(beta, X, prior_var)

Hessian of negative log-posterior for logistic regression.
"""
function _hessian_logistic(
    beta::AbstractVector{<:Real},
    X::AbstractMatrix{<:Real},
    prior_var::Real
)
    linear = X * beta
    linear = clamp.(linear, -500, 500)
    prob = _sigmoid(linear)

    # Weight matrix
    W = prob .* (1 .- prob)

    # Hessian = X' diag(W) X + I/prior_var
    n_coef = size(X, 2)
    hessian = X' * Diagonal(W) * X + I(n_coef) / prior_var

    return hessian
end


"""
    bayesian_propensity_logistic(treatment, covariates; kwargs...)

Bayesian propensity score estimation using logistic regression with Laplace approximation.

# Arguments
- `treatment`: Binary treatment indicator (0/1)
- `covariates`: Covariate matrix

# Keyword Arguments
- `prior_sd=10.0`: Prior SD for coefficients
- `n_posterior_samples=1000`: Number of posterior samples
- `include_intercept=true`: Whether to include intercept

# Returns
- `BayesianPropensityResult`
"""
function bayesian_propensity_logistic(
    treatment::AbstractVector{<:Real},
    covariates::AbstractMatrix{<:Real};
    prior_sd::Real = 10.0,
    n_posterior_samples::Int = 1000,
    include_intercept::Bool = true
)
    T = Float64.(treatment)
    n = length(T)

    if size(covariates, 1) != n
        throw(ArgumentError(
            "Length mismatch: treatment ($n) != covariates ($(size(covariates, 1)))"
        ))
    end

    if !all(t -> t == 0 || t == 1, T)
        throw(ArgumentError("Treatment must be binary (0 or 1)"))
    end

    if prior_sd <= 0
        throw(ArgumentError("prior_sd must be positive, got $prior_sd"))
    end

    # Add intercept if requested
    X = if include_intercept
        hcat(ones(n), Float64.(covariates))
    else
        Float64.(covariates)
    end

    n_coef = size(X, 2)
    prior_var = prior_sd^2

    # Find MAP estimate using simple gradient descent
    beta = zeros(n_coef)
    lr = 0.01
    for _ in 1:1000
        linear = X * beta
        linear = clamp.(linear, -500, 500)
        prob = _sigmoid(linear)

        # Gradient
        grad_ll = X' * (T .- prob)
        grad_prior = -beta / prior_var
        grad = grad_ll + grad_prior

        beta += lr * grad

        # Check convergence
        if maximum(abs.(grad)) < 1e-6
            break
        end
    end
    beta_map = beta

    # Compute Hessian at MAP for Laplace approximation
    hessian = _hessian_logistic(beta_map, X, prior_var)
    cov_matrix = try
        inv(hessian)
    catch
        pinv(hessian)
    end

    # Make sure covariance is positive definite
    cov_matrix = (cov_matrix + cov_matrix') / 2
    eigvals_cov = eigvals(cov_matrix)
    if minimum(eigvals_cov) < 0
        cov_matrix = cov_matrix - minimum(eigvals_cov) * I + 1e-6 * I
    end

    # Draw coefficient samples from approximate posterior
    beta_samples = try
        rand(MvNormal(beta_map, cov_matrix), n_posterior_samples)'
    catch
        stds = sqrt.(diag(cov_matrix))
        randn(n_posterior_samples, n_coef) .* stds' .+ beta_map'
    end

    # Compute propensity samples
    linear_samples = beta_samples * X'  # (n_samples, n)
    propensity_samples = _sigmoid.(linear_samples)

    # Summary statistics
    posterior_mean = vec(mean(propensity_samples, dims=1))
    posterior_sd = vec(std(propensity_samples, dims=1))

    coefficient_mean = vec(mean(beta_samples, dims=1))
    coefficient_sd = vec(std(beta_samples, dims=1))

    mean_uncertainty = mean(posterior_sd)
    propensity_range = maximum(posterior_mean) - minimum(posterior_mean)

    return BayesianPropensityResult(
        propensity_samples,
        posterior_mean,
        posterior_sd,
        nothing,
        0,
        nothing,
        0.0,
        0.0,
        "logistic_laplace",
        n,
        Int(sum(T)),
        Int(n - sum(T)),
        mean_uncertainty,
        propensity_range,
        coefficient_mean,
        coefficient_sd,
        beta_samples,
        prior_sd
    )
end


# Handle vector covariates
function bayesian_propensity_stratified(
    treatment::AbstractVector{<:Real},
    covariates::AbstractVector{<:Real};
    kwargs...
)
    return bayesian_propensity_stratified(treatment, reshape(covariates, :, 1); kwargs...)
end

function bayesian_propensity_logistic(
    treatment::AbstractVector{<:Real},
    covariates::AbstractVector{<:Real};
    kwargs...
)
    return bayesian_propensity_logistic(treatment, reshape(covariates, :, 1); kwargs...)
end


"""
    bayesian_propensity(treatment, covariates; method="auto", kwargs...)

Bayesian propensity score estimation with automatic method selection.

# Arguments
- `treatment`: Binary treatment indicator (0/1)
- `covariates`: Covariate matrix
- `method`: "auto", "stratified", or "logistic"
"""
function bayesian_propensity(
    treatment::AbstractVector{<:Real},
    covariates::AbstractMatrix{<:Real};
    method::String = "auto",
    kwargs...
)
    if method == "auto"
        # Heuristic: use stratified if few unique values per covariate
        n_unique = [length(unique(covariates[:, j])) for j in 1:size(covariates, 2)]
        if all(nu -> nu <= 10, n_unique)
            method = "stratified"
        else
            method = "logistic"
        end
    end

    # Filter kwargs to only include parameters accepted by the selected method
    stratified_params = (:prior_alpha, :prior_beta, :n_bins, :n_posterior_samples)
    logistic_params = (:prior_sd, :n_posterior_samples, :include_intercept)

    if method == "stratified"
        filtered_kwargs = filter(kv -> kv.first in stratified_params, kwargs)
        return bayesian_propensity_stratified(treatment, covariates; filtered_kwargs...)
    elseif method == "logistic"
        filtered_kwargs = filter(kv -> kv.first in logistic_params, kwargs)
        return bayesian_propensity_logistic(treatment, covariates; filtered_kwargs...)
    else
        throw(ArgumentError("Unknown method: $method. Use 'auto', 'stratified', or 'logistic'."))
    end
end

function bayesian_propensity(
    treatment::AbstractVector{<:Real},
    covariates::AbstractVector{<:Real};
    kwargs...
)
    return bayesian_propensity(treatment, reshape(covariates, :, 1); kwargs...)
end
